src/lib/pickles/marlin_checks/marlin_checks.ml

```diff
@@ -0,0 +1,138 @@
+
+  val ( * ) : t -> t -> t
+
+  val ( / ) : t -> t -> t
+
+  val ( + ) : t -> t -> t
+
+  val ( - ) : t -> t -> t
+end
+
+type 'f field = (module Field_intf with type t = 'f)
+
+let map_reduce reduce xs map = List.reduce_exn (List.map xs ~f:map) ~f:reduce
+
+(* x^{2 ^ k} - 1 *)
+let vanishing_polynomial (type t) ((module F) : t field) domain x =
+  let k = Domain.log2_size domain in
+  let rec pow2pow acc i =
+    if i = 0 then acc else pow2pow F.(acc * acc) (i - 1)
+  in
+  F.(pow2pow x k - one)
+
+let domain (type t) ((module F) : t field) (domain : Domain.t) : t domain =
+  let size = F.of_int (Domain.size domain) in
+  object
+    method size = size
+
+    method vanishing_polynomial x = vanishing_polynomial (module F) domain x
+  end
+
+let all_but m = List.filter Abc.Label.all ~f:(( <> ) m)
+
+let checks (type t) (module F : Field_intf with type t = t) ~input_domain
+    ~domain_h ~domain_k ~x_hat_beta_1
+    { Composition_types.Dlog_based.Proof_state.Deferred_values.Marlin.sigma_2
+    ; sigma_3
+    ; alpha
+    ; eta_a
+    ; eta_b
+    ; eta_c
+    ; beta_1
+    ; beta_2
+    ; beta_3 }
+    { Pairing_marlin_types.Evals.w_hat
+    ; z_hat_a
+    ; z_hat_b
+    ; g_1
+    ; h_1
+    ; g_2
+    ; h_2
+    ; g_3
+    ; h_3
+    ; row= {a= row_a; b= row_b; c= row_c}
+    ; col= {a= col_a; b= col_b; c= col_c}
+    ; value= {a= value_a; b= value_b; c= value_c}
+    ; rc= {a= rc_a; b= rc_b; c= rc_c} } =
+  let open F in
+  let abc = Abc.abc in
+  (* Marlin checks follow *)
+  let row = abc row_a row_b row_c
+  and col = abc col_a col_b col_c
+  and value = abc value_a value_b value_c
+  and rc = abc rc_a rc_b rc_c
+  and eta = abc eta_a eta_b eta_c
+  and z_ = abc z_hat_a z_hat_b (z_hat_a * z_hat_b) in
+  let z_hat =
+    let v_X_beta_1 = input_domain#vanishing_polynomial beta_1 in
+    (w_hat * v_X_beta_1) + x_hat_beta_1
+  in
+  let sum = map_reduce ( + ) in
+  let prod = map_reduce ( * ) in
+  let r_alpha =
+    let v_h_alpha = domain_h#vanishing_polynomial alpha in
+    fun x -> (v_h_alpha - domain_h#vanishing_polynomial x) / (alpha - x)
+  in
+  let v_h_beta_1 = domain_h#vanishing_polynomial beta_1 in
+  let v_h_beta_2 = domain_h#vanishing_polynomial beta_2 in
+  let a_beta_3, b_beta_3 =
+    let beta_1_beta_2 = beta_1 * beta_2 in
+    let term =
+      Memo.general ~cache_size_bound:10 (fun m ->
+          beta_1_beta_2 + rc m - (beta_1 * row m) - (beta_2 * col m) )
+    in
+    let a =
+      v_h_beta_1 * v_h_beta_2
+      * sum Abc.Label.all (fun m -> eta m * value m * prod (all_but m) term)
+    in
+    let b = prod Abc.Label.all term in
+    (a, b)
+  in
+  [ ( h_3 * domain_k#vanishing_polynomial beta_3
+    , a_beta_3 - (b_beta_3 * ((beta_3 * g_3) + sigma_3)) )
+  ; ( r_alpha beta_2 * sigma_3 * domain_k#size
+    , (h_2 * v_h_beta_2) + sigma_2 + (g_2 * beta_2) )
+  ; ( r_alpha beta_1 * sum Abc.Label.all (fun m -> eta m * z_ m)
+    , (h_1 * v_h_beta_1) + (beta_1 * g_1) + (sigma_2 * domain_h#size * z_hat)
+    ) ]
```

Comment:
Would be handy to add a link to the paper and mention page 32 in a code comment here.

I think I may be misunderstanding how the domain sizes should be distributed. or where they are already factored in. From the paper, I would have expected this to be

```ocaml
[ ( h_3 * domain_k#vanishing_polynomial beta_3 * domain_k#size
  , a_beta_3 * domain_k#size - (b_beta_3 * ((beta_3 * g_3) * domain_k#size + sigma_3)) )
; ( r_alpha beta_2 * sigma_3 * domain_h#size
  , ((h_2 * v_h_beta_2) + (beta_2 * g_2)) * domain_h#size + sigma_2 )
; ( (s beta_1 (* [s = 0]? *) + (r_alpha beta_1 * sum Abc.Label.all (fun m -> eta m * z_ m)) - (sigma_2 * z_hat)) * domain_h#size
  , ((h_1 * v_h_beta_1) + (beta_1 * g_1)) * domain_h#size + sigma_1 ) ]
```

If there is a different intuition to use here, it might be good to pop that in a comment here too.

I presume that the final check was elided because it was redundant, along the lines of
```ocaml
(z_hat_a * z_hat_b - (z_hat_a * z_hat_b), h_0 * v_h_beta_1)
<=>
(0, h_0)
```
but it would be nice to clarify that as well.

Reply:
Indeed -- our sigma_i are equal to the paper's sigma_i / domain K or sigma_i / domain H as relevant. It is equivalent since `sigma_i` and `sigma_i / C` (where `C` is a constant) can easily be computed from each other. I can add a comment.

The final check is elided since we use an optimized variant in which `h_0` is omitted.

Reply:
I think the `* domain_k#size` might need to be `* domain_h#size` in the second check, but I'll review this against the paper again with this in mind.